File: frontend/src/app/dto/game.ts
import {GameType} from "../enums/game-type";
import {GameState} from "../enums/game-state";
import {Player} from "./player";
import {RuleSet} from "./rule-set";
import {LogEntry} from "./log-entry";
import {Vote} from "./vote";

export abstract class Game {
  protected constructor(public id: number,
                        public gameType: GameType,
                        public gameIdentifier: GameIdentifier,
                        public gameState: GameState,
                        public players: Set<Player>,
                        public ruleSet: RuleSet,
                        public round: number,
                        public votes: Vote<any, any, any>[],
                        public log: LogEntry[],
                        public startedAt: string,
                        public endedAt: string | null) {
  }
}

export class GameIdentifier {
  constructor(public token: string) {
  }
}

export class GameStartRequest {
  constructor(public gameType: GameType) {
  }
}

export class GameBaseInformation {
  constructor(public gameIdentifier: GameIdentifier,
              public gameType: GameType,
              public gameState: GameState) {
  }
}

File: backend/src/main/java/com/fschoen/parlorplace/backend/entity/Game.java
package com.fschoen.parlorplace.backend.entity;

import com.fschoen.parlorplace.backend.enumeration.GameState;
import com.fschoen.parlorplace.backend.enumeration.GameType;
import lombok.*;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.LazyCollection;
import org.hibernate.annotations.LazyCollectionOption;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import javax.persistence.*;
import javax.validation.constraints.Min;
import javax.validation.constraints.NotNull;
import java.time.Instant;
import java.util.List;
import java.util.Set;

@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder(toBuilder = true)
@Data
@Entity
@Inheritance(strategy = InheritanceType.TABLE_PER_CLASS)
public abstract class Game<P extends Player<?>, RS extends RuleSet, V extends Vote<P, ?, ?, ?, ?>, L extends LogEntry<?>> {

    @Id
    @GeneratedValue(strategy = GenerationType.AUTO, generator = "seq_game_instance_id")
    @SequenceGenerator(name = "seq_game_instance_id", sequenceName = "seq_game_instance_id")
    protected Long id;

    @OneToOne(cascade = CascadeType.ALL, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(referencedColumnName = "id")
    @NotNull
    protected GameIdentifier gameIdentifier;

    @Column(nullable = false)
    @Enumerated
    @NotNull
    protected GameState gameState;

    @OneToMany(mappedBy = "game", targetEntity = Player.class, fetch = FetchType.EAGER, cascade = CascadeType.ALL, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @EqualsAndHashCode.Exclude
    @NotNull
    protected Set<P> players;

    @OneToOne(targetEntity = RuleSet.class, cascade = CascadeType.ALL, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @JoinColumn(referencedColumnName = "id")
    @EqualsAndHashCode.Exclude
    @NotNull
    protected RS ruleSet;

    @Column(nullable = false)
    @NotNull
    @Min(0)
    protected Integer round;

    @OneToMany(mappedBy = "game", targetEntity = Vote.class, cascade = CascadeType.ALL, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @LazyCollection(LazyCollectionOption.FALSE)
    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @NotNull
    protected List<V> votes;

    @OneToMany(mappedBy = "game", targetEntity = LogEntry.class, cascade = CascadeType.ALL, orphanRemoval = true)
    @OnDelete(action = OnDeleteAction.CASCADE)
    @LazyCollection(LazyCollectionOption.FALSE)
    @EqualsAndHashCode.Exclude
    @ToString.Exclude
    @NotNull
    protected List<L> log;

    @Column(nullable = false)
    protected Instant startedAt;

    @Column
    protected Instant endedAt;

    public abstract GameType getGameType();

}
